xarray/ufuncs.py

```diff
@@ -0,0 +1,366 @@
+"""xarray specific universal functions."""
+
+import textwrap
+
+import numpy as np
+
+import xarray as xr
+from xarray.core.groupby import GroupBy
+
+
+def _walk_array_namespaces(obj, namespaces):
+    if isinstance(obj, xr.DataTree):
+        # TODO: DataTree doesn't actually support ufuncs yet
+        for node in obj.subtree:
+            _walk_array_namespaces(node.dataset, namespaces)
+    elif isinstance(obj, xr.Dataset):
+        for name in obj.data_vars:
+            _walk_array_namespaces(obj[name], namespaces)
+    elif isinstance(obj, GroupBy):
+        _walk_array_namespaces(next(iter(obj))[1], namespaces)
+    elif isinstance(obj, xr.DataArray | xr.Variable):
+        _walk_array_namespaces(obj.data, namespaces)
+    else:
+        namespace = getattr(obj, "__array_namespace__", None)
+        if namespace is not None:
+            namespaces.add(namespace())
+
+    return namespaces
+
+
+def get_array_namespace(*args):
+    xps = set()
+    for arg in args:
+        _walk_array_namespaces(arg, xps)
+
+    xps.discard(np)
+    if len(xps) > 1:
+        names = [module.__name__ for module in xps]
+        raise ValueError(f"Mixed array types {names} are not supported.")
+
+    return next(iter(xps)) if len(xps) else np
+
+
+class _UnaryUfunc:
+    """Wrapper for dispatching unary ufuncs."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, x, **kwargs):
```

Comment:
This is a good place to use positional only arguments:
```suggestion
    def __call__(self, x, /, **kwargs):
```